src/serde_impl.rs

```diff
@@ -3,16 +3,169 @@
+        fn serialize_secret<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error>;
+    }
+
+    impl<T: SerializeSecret> SerializeSecret for &T {
+        fn serialize_secret<S: Serializer>(
+            &self,
+            serializer: S,
+        ) -> Result<<S as Serializer>::Ok, <S as Serializer>::Error> {
+            (*self).serialize_secret(serializer)
+        }
+    }
+}
+
+/// `SerdeSecret` is a wrapper struct for serializing and deserializing secret keys. Due to security
+/// concerns serialize shouldn't be implemented for secret keys to avoid accidental leakage.
+///
+/// Whenever this struct is used the integrity of security boundaries should be checked carefully.
+pub struct SerdeSecret<T>(T);
+
+impl<T> Deref for SerdeSecret<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        self.inner()
+    }
+}
+
+impl<T> SerdeSecret<T> {
+    /// Returns the actual secret from the wrapper
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+
+    /// Returns a reference to the actual secret contained in the wrapper
+    pub fn inner(&self) -> &T {
+        &self.0
+    }
+}
+
+impl<'de, T: Deserialize<'de>> Deserialize<'de> for SerdeSecret<T> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Ok(SerdeSecret(Deserialize::deserialize(deserializer)?))
+    }
+}
+
+impl<'de, T: SerializeSecret> Serialize for SerdeSecret<T> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.0.serialize_secret(serializer)
+    }
+}
+
+#[cfg(not(feature = "use-insecure-test-only-mock-crypto"))]
+impl<'de> Deserialize<'de> for crate::SecretKey {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        use pairing::bls12_381;
+        use pairing::PrimeField;
+
+        use serde::de;
+        use std::fmt;
+
+        struct ReprVisitor;
+        impl<'de> serde::de::Visitor<'de> for ReprVisitor {
+            type Value = [u64; 4];
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                write!(formatter, "A tuple of four u64 integers.")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: de::SeqAccess<'de>,
+            {
+                let mut repr = [0u64; 4];
+                for (num, part) in repr.iter_mut().enumerate() {
+                    *part = match seq.next_element()? {
+                        Some(x) => x,
+                        None => return Err(de::Error::invalid_length(num + 1, &"Expected 4 u64s")),
+                    };
+                }
+                Ok(repr)
+            }
+        }
+
+        let repr = deserializer.deserialize_tuple(4, ReprVisitor)?;
+        let mut fr = match bls12_381::Fr::from_repr(bls12_381::FrRepr(repr)) {
```

Comment:
Isn't all of the above equivalent to just deserializing the `[u64; 4]` directly here?
```rust
Fr::from_repr(FrRepr(Deserialize::deserialize(deserializer)?)
```
And maybe that would even work with mock crypto, since the type to deserialize is just inferred?

Reply:
Oh, you are right, I assumed that `Deserialize` wasn't implemented for arrays, but it is for small ones like this.